plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReLoader.java

```diff
@@ -0,0 +1,285 @@
+        if (!enableAutoReLoadModel) {
+            return;
+        }
+
+        // At opensearch startup, get local node id, if not ml node,we ignored, just return without doing anything
+        String localNodeId = clusterService.localNode().getId();
+        if (!MLNodeUtils.isMLNode(nodeHelper.getNode(localNodeId))) {
+            return;
+        }
+
+        // auto reload all models of this local node, if it fails, reTryTimes+1, if it succeeds, reTryTimes is cleared to 0
+        threadPool.generic().submit(() -> autoReLoadModelByNodeId(localNodeId));
+    }
+
+    /**
+     * auto reload all the models under the node id
+     * the node must be a ml node
+     * @param localNodeId node id
+     */
+    @VisibleForTesting
+    void autoReLoadModelByNodeId(String localNodeId) {
+        if (!isExistedIndex(ML_TASK_INDEX)) {
+            return;
+        }
+
+        SearchResponse response = queryTask(localNodeId);
+        if (response == null || response.getHits() == null) {
+            return;
+        }
+
+        SearchHit[] hits = response.getHits().getHits();
+        if (CollectionUtils.isEmpty(hits)) {
+            return;
+        }
+
+        // According to the node id to get retry times, if more than the maxi retry times, don't need to retry
+        // that the number of unsuccessful reload has reached the maximum number of times, do not need to reload
+        int reTryTimes = getReTryTimes(localNodeId);
+        if (reTryTimes > ML_MODEL_RELOAD_MAX_RETRY_TIMES) {
+            log.info("have exceeded max retry times, always failure");
+            return;
+        }
+
+        try (XContentParser parser = createXContentParserFromRegistry(xContentRegistry, hits[0].getSourceRef())) {
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+            MLTask mlTask = MLTask.parse(parser);
+
+            autoReLoadModelByNodeAndModelId(localNodeId, mlTask.getModelId());
+
+            // if reload the model successfully,the number of unsuccessful reload should be reset to zero.
+            reTryTimes = 0;
+        } catch (RuntimeException | IOException e) {
+            reTryTimes++;
+            log.error("Can't auto reload model in node id {} ,has tried {} times\nThe reason is:{}", localNodeId, reTryTimes, e);
+        }
+
+        // Store the latest value of the reTryTimes and node id under the index ".plugins-ml-model-reload"
+        saveLatestReTryTimes(localNodeId, reTryTimes);
+    }
+
+    /**
+     *  auto reload 1 model under the node id
+     * @param localNodeId node id
+     * @param modelId model id
+     */
+    @VisibleForTesting
+    void autoReLoadModelByNodeAndModelId(String localNodeId, String modelId) throws IllegalArgumentException {
+        MLLoadModelRequest mlLoadModelRequest = new MLLoadModelRequest(modelId, new String[] { localNodeId }, false, false, true);
+        AtomicReference<LoadModelResponse> loadModelResponse = new AtomicReference<>();
```

Comment:
Why do we need `AtomicReference` instead of `Supplier`, this method doesn't seems to be race conditioned.